### DISCUSSION TITLE
Writing gltf model

### DISCUSSION THREAD

**Discussion by raducal** [donmccurdy/glTF-Transform#552] [Q&A] [answered]
Hey,
I'm trying to create a gltf model from scratch using vertices and face indices I calculate but the model isn't coming out as expected. 

Here is what I'm expecting vs what is being created
Expected
![good](https://user-images.githubusercontent.com/55283556/162946864-33629c57-5821-4938-95d3-75c0e5ba4ea3.PNG)

What I'm getting
![bad](https://user-images.githubusercontent.com/55283556/162946921-97051a5d-a669-4bcd-8d1f-70df0e415e32.PNG)

How I'm creating the model
```
const createGLTFModel = async (vertices: number[], faces: number[]) => {
  const doc = new Document();

  const buffer = doc.createBuffer();

  const position = doc
    .createAccessor()
    .setArray(new Float32Array(vertices))
    .setType("VEC3")
    .setBuffer(buffer);
  const indices = doc
    .createAccessor()
    .setArray(new Uint16Array(faces))
    .setType("SCALAR")
    .setBuffer(buffer);

  const prim = doc
    .createPrimitive()
    .setAttribute("POSITION", position)
    .setIndices(indices);

  const mesh = doc.createMesh("mesh").addPrimitive(prim);
  const node = doc.createNode("node").setMesh(mesh);
  doc.createScene().addChild(node);

  const glb = await new WebIO().writeBinary(doc);

  return new Blob([glb.buffer], { type: "application/object" });
};
```

Small sample of my vertices + face indices
Vertices
1 1 14
1 2 14
2 2 14
2 1 14
1 1 15
1 2 15
2 2 15
2 1 15

Face Indices
1 2 3
1 3 4
8 7 6
8 6 5
4 3 7
4 7 8
5 1 4
5 4 8
5 6 2
5 2 1
2 6 7
2 7 3

Vertices and face indices are being stored in separate 1d arrays. These are also being used to create obj and stl files and they're working fine for those. Any help would be much appreciated

**Comment by raducal** [comment, marked as answer]
Ok so it seems its because my face indices are 1 indexed but they should be 0 indexed. Changing that fixed the issue for smaller models but larger models still seem to be broken

EDIT: Changing face indices to Uint32Array fixes the larger model issue as well
```
  const indices = doc
    .createAccessor()
    .setArray(new Uint32Array(faces))
    .setType("SCALAR")
    .setBuffer(buffer);
```

**Reply by donmccurdy** [reply]
Glad you found the answer! The other thing I was going to mention was winding order - the [winding order of triangles must be counter-clockwise](https://www.khronos.org/registry/glTF/specs/2.0/glTF-2.0.html#instantiation). If your mesh is looking correct now then it's probably got the right winding order anyway. :)
